doc: generated API documentation using sphinx autoapi

sankha555 opened the pull request:
* [x] fix #186 
* [x] used sphinx autoapi to generate API documentation for the repository ([tutorial](https://sphinx-autoapi.readthedocs.io/en/latest/tutorials.html#setting-up-automatic-api-documentation-generation))
* [x] tests passed

sankha555 commented:
@matthiaskoenig I have made a PR to add the API documentation built using the Sphinx autoapi extension. The documentation has been build in the ```docs_builder/_build``` directory. To add the directory in the push, I commented out the corresponding line in the ```.gitignore``` file. The homepage of the docs is present in ```docs_builder/_build/index.html```.

I was facing one error during the build. The build always failed on reading the ```/src/sbmlutils/resources/converters/odefac_template.py``` file. It generated the following error:

![image](https://user-images.githubusercontent.com/54751139/107770379-4e6cb580-6d5f-11eb-81cc-c19169766c99.png)

I think the builder was not able to parse these and similar constructs in the file (I have seen such constructs in Jinja templates too) :
![image](https://user-images.githubusercontent.com/54751139/107770574-a1466d00-6d5f-11eb-8409-01843500d62c.png)

As a test, I commented out the entire file and ran the build again and this time the build was successful. The entire documentation was generated under 2 minutes (snapshot attached below). After that, I restored the ```odefac_template.py``` file to its original version. But I am afraid that this might not be the correct method to build documentation. I wanted to ask your suggestions as to how should I resolve the issue and if I should rebuild the documentation after doing required changes.

![image](https://user-images.githubusercontent.com/54751139/107771042-334e7580-6d60-11eb-8071-1348dc55eff6.png)

Please guide me as to how should I proceed further with this pull request and I would make the required changes to the patch. Thanks!

matthiaskoenig commented:
You have to exclude the template files from the api autodocumentation. This can be configured. Have a look at the autoapi documentation how to exclude files and update this pull request.

sankha555 commented:
@matthiaskoenig I added the configuration to exclude the template files ```autoapi_ignore = ["*template*"]``` to the ```conf.py``` file and it worked successfully, thanks a lot for the clue. Please take a look at the patch and please let me know if I should make any further changes. Thanks!

matthiaskoenig commented on .gitignore:
```diff
@@ -32,6 +32,6 @@ docs_builder/notebooks/.ipynb_checkpoints/
 misc/biomodels/r31_uncurated/
 
 # documentation build
-docs_builder/_build/
```
This has to be reverted. The build files do NOT belong in the repository. These are created on the fly via readthedocs. I.e. revert the changes and remove all the `_docs_build/_build` files from the pull request.

sankha555 commented:
Okay so to create the API documentation, only the information in the ```conf.py``` is important and we don't need to keep the build files in the repository too. readthedocs then just creates the documentation from the data in the configuration file. Have I understood the situation correctly? 

Sure, I will remove the ```_build``` directory from the PR and also restore the ```.gitignore```. Thanks!

sankha555 commented:
I have made removed all the files from the ```docs_build/_build/``` directory. Now, only these files remain in the PR:

- ```docs_build/_build/conf.py``` :  added configurations for api autodocumentation and after running isort and black
- ```docs_build/requirements-docs.txt``` : added dependency ```sphinx-autoapi``` for api autodocumentation

The ```.gitignore``` has also been restored to always ignore ```_build``` files.

Please let me know if I should make any further changes to the pull request. Thanks!

matthiaskoenig commented on docs_builder/requirements-docs.txt:
```diff
@@ -4,4 +4,4 @@ nbsphinx>=0.8.1
 pypandoc>=1.4
 jupyterlab>=3.0.5
-
+sphinx-autoapi>=1.7.0
```
always use the latest dependencies. We try to enforce the latest versions as much as possible.
I will update this.

sankha555 replied:
I understand the situation, I think I had just added the sphinx-autoapi version (i got it from [pypi.org](https://pypi.org/project/sphinx-autoapi/) webpage) and it had slipped my mind to update the remaining dependencies. I hope this is what we are looking to update here. If you say, I can make an update for the all the dependencies:

- ```pypandoc``` to 1.5
- ```jupyterlab``` to 3.0.7

 The latest version of ```sphinx-autoapi``` is shown as 1.7.0, should we change it to some other version? Shall I go ahead and make these changes to the dependencies?
